# Computação Científica 1 - 2023.2 - Prof. Adriano Côrtes

## Teste 1

***

### Questão 1
Explique o comportamento (surpreendente?) do código abaixo

In [2]:
d = 8
e = 2
from math import *
sqrt(d ** e)

16.88210319127114

O motivo é simples, ao realizarmos a importação da biblioteca math através do seguinte trecho from math import *, estaremos importando também "e" como sendo o número de euler dado por 2.71828... e com isso ao realizarmos o cálculo sqrt(d**e) obteremos aquele valor 16.88210319127114. Com isso vemos que é importante saber o que estamos importando ao rodarmos o código acima

### Questão 2
Como sabemos números reais são representados no computador por números de precisão finita, os números em ponto-flutuante. No Python, assim como na maioria das outras linguagens de programação predomina o padrão IEEE-754, em particular a representação em precisão dupla (*double precision*) é o default em Python. Podemos investigar essas informações por meio do módulo `sys`. Essas informações estão atributo `sys.float_info`:

In [3]:
import sys
print(sys.float_info)
print('O menor número representável em ponto-flutuante de precisão dupla é {}'.format(sys.float_info.min))
print('O maior número representável em ponto-flutuante de precisão dupla é {}'.format(sys.float_info.max))

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)
O menor número representável em ponto-flutuante de precisão dupla é 2.2250738585072014e-308
O maior número representável em ponto-flutuante de precisão dupla é 1.7976931348623157e+308


1. O que acontece se você tentar avaliar a expressão $e^{1000}$ que gera um número maior que o maior número representável?
2. Como você calcularia a hipotenusa $c$ de um triângulo retângulo cujos catetos são $a = 1.5 \times 10^{200}$ e $b = 3.5 \times 10^{201}$? **Note que calcular $a^2$ e $b^2$ não é uma opção, como no item anterior**. Verifique seu resultado usando a função `math.hypot`

Ao tentarmos avaliar $e^{1000}$ geramos um número muito grande, maior até que o maior número representável em ponto flutuante. Retornando um OverflowError.

In [4]:
e**1000

OverflowError: (34, 'Result too large')

Para calcular a hipotenusa $c$ do triângulo retângulo, usaremos a função `math.hypot`.

In [5]:
import math
a = 1.5e+200
b = 3.5e+201
c = math.hypot(a,b)
print(c)

3.5032128111206724e+201


### Questão 3
Usando uma tupla de strings com o nome dos dígitos de $0$ a $9$, crie um trecho de código em Python que imprima até a sétima casas decimais de $\pi$ na forma de texto, como:

`três ponto um quatro um cinco nove dois seis`

***

In [13]:
numeros = ("zero", "um", "dois", "três", "quatro", "cinco", "seis", "sete", "oito", "nove")
pi_str = "{:.7f}".format(math.pi)
for unit in pi_str:
    if unit == ".":
        print("ponto", end=" ")
    else:
        print(numeros[int(unit)], end=" ")

três ponto um quatro um cinco nove dois sete 

In [7]:
print(math.pi)

3.141592653589793


Perceba que ao utilizarmos a função `math.pi` e fizermos o `{:.7f}.format(math.pi)`, obtemos um arredondamento no sétimo dígito que ao invés de $6$ fica $7$, para alterarmos isso podemos definir $\pi$ como sendo $3.1415926$

In [14]:
numeros = ("zero", "um", "dois", "três", "quatro", "cinco", "seis", "sete", "oito", "nove")
pi_str = "{:.7f}".format(3.1415926)
for unit in pi_str:
    if unit == ".":
        print("ponto", end=" ")
    else:
        print(numeros[int(unit)], end=" ")

três ponto um quatro um cinco nove dois seis 

### Questão 4
Escreva um trecho que código Python que imprima de maneira formatada (bonitinha) as oito primeiro linhas do triângulo de Pascal.

In [76]:
from math import factorial
n = 8 #fiquemos atento que o n neste caso é o número de linhas
for i in range(n):
    for j in range(n-i):
        print(end=" ")
    for j in range(i+1):
        print(factorial(i)//(factorial(j)*factorial(i-j)), end=" ")
    print()

        1 
       1 1 
      1 2 1 
     1 3 3 1 
    1 4 6 4 1 
   1 5 10 10 5 1 
  1 6 15 20 15 6 1 
 1 7 21 35 35 21 7 1 


Para melhor visualização dependendo do objetivo, podemos fazer da seguinte forma:
|||||||||
|-----|--------|--------|--------|--------|--------|--------|--------|
|1|
|1    | 1  |
|1    | 2  | 1  |
|1    | 3  | 3  | 1  |
|1    | 4  | 6  | 4  | 1  |
|1    | 5  | 10  | 10  | 5  | 1  |
|1    | 6  | 15  | 20  | 15  | 6  | 1  |
|1    | 7  | 21  | 35  | 35  | 21  | 7  | 1  |

### Questão 5
1. Crie uma função que receba como entrada um número inteiro $N$ e retorna uma lista com os primeiros $N$ números de Fibonacci.
2. A chamada [*lei de Benford*](https://pt.wikipedia.org/wiki/Lei_de_Benford) é uma lei baseada na observação sobre a frequência do primeiro dígito $d \in \{1,2,\ldots,9\}$ de uma determinada coleção de números (*data sets*). Segundo essa lei a distribuição logaritmica, portanto não-uniforme, dada por $$P(d) = \log_{10} \left( \dfrac{d+1}{d} \right),$$ ou seja, número começando com $1$ são mais frequentes que os começando com $2$, e assim em diante, com efeito temos aproximadamente

| $d$ | $P(d)$ |
|-----|--------|
|1    | 0.301  |
|2    | 0.176  |
|3    | 0.125  |
|4    | 0.097  |
|5    | 0.079  |
|6    | 0.067  |
|7    | 0.058  |
|8    | 0.051  |
|9    | 0.046  |

Usando a função criada no item anterior calcule a distribuição dos primeiros dígitos da sequência de Fibonacci para $N=500$ e $N=1000$, e avalie se ela segue a lei de Benford.

### Primeira parte da questão

In [61]:
# Começamos a questão definindo a sequência de fibonacci como uma função
def Fibonacci(N):
    if N <= 0:
        return []
    elif N == 1:
        return [0]
    elif N == 2:
        return [0,1]
    else:
        fib_seq = [0,1]
        while len(fib_seq) <= N:
            new_term = fib_seq[-1] + fib_seq[-2]
            fib_seq.append(new_term)
    return fib_seq

### Segunda parte da questão

In [63]:
# Agora iremos calcular as proporções para N = 500
fib_seq = Fibonacci(500)
first_seq = [0,0,0,0,0,0,0,0,0]
for num in fib_seq:
    first_num = int(str(num)[0])
    if first_num != 0:
        first_seq[first_num-1] += 1
    
total = sum(first_seq)
proportion = []
for first in first_seq:
    proportion_count = first/total
    proportion.append(proportion_count)
print(proportion)

[0.302, 0.176, 0.126, 0.094, 0.08, 0.066, 0.058, 0.054, 0.044]


Temos então para N = 500:

| First Num | Proportion |
|-----|--------|
|1    | 0.302  |
|2    | 0.176  |
|3    | 0.126  |
|4    | 0.094  |
|5    | 0.08  |
|6    | 0.066  |
|7    | 0.058  |
|8    | 0.054  |
|9    | 0.044  |

In [65]:
# Agora iremos calcular as proporções para N = 1000
fib_seq = Fibonacci(1000)
first_seq = [0,0,0,0,0,0,0,0,0]
for num in fib_seq:
    first_num = int(str(num)[0])
    if first_num != 0:
        first_seq[first_num-1] += 1
    
total = sum(first_seq)
proportion = []
for first in first_seq:
    proportion_count = first/total
    proportion.append(proportion_count)
print(proportion)

[0.301, 0.177, 0.125, 0.096, 0.08, 0.067, 0.056, 0.053, 0.045]


Temos então para N = 1000:

| First Num | Proportion |
|-----|--------|
|1    | 0.301  |
|2    | 0.177  |
|3    | 0.125  |
|4    | 0.096  |
|5    | 0.08  |
|6    | 0.067  |
|7    | 0.056  |
|8    | 0.053  |
|9    | 0.045  |

In [66]:
# Vale lembrar que os valores esperados segundo a lei de Benford são:
benford = [0.301,0.176,0.125,0.097,0.079,0.067, 0.058, 0.051, 0.046]


Como os valores calculados se aproximam muito dos valores esperados pela Lei de Benford, podemos dizer que a sequência de Fibonacci segue a lei para os primeiros dígitos.

### Questão 6
Escreva um loop usando `while` para calcular a [média aritmética-geometrica](https://en.wikipedia.org/wiki/Arithmetic%E2%80%93geometric_mean) (AGM) de dois números reais positivos, $x$ e $y$, definida como o limite das sequência:

$$ 
\begin{align}
a_{n+1} &= \frac{1}{2} (a_n + g_n), \\
g_{n+1} &= \sqrt{a_n g_n},
\end{align}
$$

começando com $a_0 = x$ e $g_0 = y$. Ambas as sequências convergem para o mesmo número, denotado por $\textrm{agm}(x,y)$. Use sua implementação para calcular a constante de *Gauss*, $G = 1/\textrm{agm}(1,\sqrt{2})$.

**Importante:** Você terá que usar um critério de parada para iteração. 

In [59]:
# Vamos começar definindo a função da média aritmética-geométrica
def agm(x,y,tol=1e-16):
    a_n = x
    g_n = y
    while abs(a_n - g_n) > tol:
        a_n, g_n = (a_n + g_n)/2, math.sqrt(a_n * g_n)
        
    return g_n

In [67]:
# Agora calcularemos a constante de gauss
G = 1/agm(1,math.sqrt(2))
print(G)

0.8346268416740731
